\documentclass[10pt]{article}
\usepackage[margin=1.0in]{geometry}
\usepackage[utf8]{inputenc}
\usepackage[parfill]{parskip}
\SweaveOpts{keep.source=TRUE} 

\title{oligoMask Vignette}
\author{Daniel Bottomly}
%\VignetteIndexEntry{oligoMask}
\begin{document}

\maketitle

\section{Introduction}

It has been observed that for RNA from a given sample, genetic differences relative to the probe sequences can affect hybridization to short oligonucleotide probes resulting in false positives or negatives depending on the experimental and array design (Walter et al. 2007; Alberts et al. 2007).
Although R packages exist to effectively deal with a two group comparison between several strains or species through procedures based mainly on the expression data (for instance maskBAD and SNEP), it is also useful to have methods to allow the a priori removal of probes which are likely to be
affected by hybridization issues (Benovoy et al. 2008).  Removing such variants ahead of time is useful when analyzing more complicated experimental designs such as those found in eQTL analyses using more genetically diverse mouse lines such as Diversity Outbred (Svenson et al. 2012),
Collaborative Cross (Collaborative Cross Consortium 2012) or other Heterogenous Stock (Chia et al. 2005) mice.  In these crosses, a set of inbred strains were used as the founders and the resulting genomes for mapping would consist of a mixture of the original founder genomes and therefore
without additional genotyping it would not be known whether a given animal possesses a given variant.  The simplest approach to variant masking in this context is to remove probes which overlap variants in any of the non-reference mouse strains prior to summarization or analysis.
The key to this is the relatively recent availability of genome-wide variant databases in VCF format such as those from the Sanger Mouse Genomes Project (http://www.sanger.ac.uk/resources/mouse/genomes/) and the 1000 Genomes for humans (http://www.1000genomes.org/).     

\section{The oligoMask package}

The \texttt{oligoMask} package is designed to work in conjunction with the \texttt{oligo} package to facilitate removal of aberrant probe expression prior to RMA procedure for Affymetrix arrays.
Currently it works by removing suspect probes from the overall expression matrix prior to the call to the RMA function.  These suspect probe annotations are most easily derived from VCF files with the parsed data stored in an SQLite database.
This database can be optionally wrapped in an R package with appropriate metadata to facilitate sharing and reproducibility.  As our main use case involves mice and specifically the Affymetrix Mouse Gene ST array, we provide convenience functions
to enable creation of an oligoMask database using VCF files from the Sanger Mouse Genomes Project and the tab-delimited probe sequence files distributed by Affymetrix as well as standard Bioconductor functionality.
As a demonstration, we will utilize a portion of the array data of (Sun et al. 2012) and a variant mask generated for the NOD/ShiLtJ mouse inbred strain from the Sanger Mouse Genomes Project. 


The expression data can be retrieved from GEO using the accession number: GSE33822.  Retrieval from R can easily be carried out using the \texttt{GEOquery} package.  As our interest was in the CEL files we first downloaded them and then
kept only the 8 arrays which corresponded to wholebrain with the vehicle treatment that were run on version 1 of the Mouse Gene ST array.

<<eval=FALSE, echo=TRUE>>=

library(GEOquery)
library(GEOmetadb)
library(pd.mogene.1.0.st.v1)

gene.files <- getGEOSuppFiles("GSE33822")

cel.dir <- tempfile()
untar(rownames(gene.files)[1], exdir=cel.dir)

meta.file <- getSQLiteFile()

meta.con <- dbConnect(SQLite(),meta.file)

annot <- dbGetQuery(meta.con, "SELECT gsm.title AS sample_name, gsm.supplementary_file AS cel_name 
	FROM gsm JOIN gpl USING (gpl) WHERE series_id = 'GSE33822' AND gpl.title LIKE
	'[MoGene-1_0-st]%' AND gsm.title LIKE 'Wholebrain_%_Veh%'")

annot$clean.cel.path <- file.path(cel.dir, basename(annot$cel_name))

dbDisconnect(meta.con)

sun.gene.fs <- read.celfiles(filenames=annot$clean.cel.path,  sampleNames=annot$sample_name,
	pkgname="pd.mogene.1.0.st.v1")

@

Here we can build a database package containing the variants from the NOD strain relative to the reference genome which is from C57BL/6J, the strain used as a comparison here.  Note that for
consistency we built a custom BSGenome package for GRCm38 from the genome provided by Sanger as their reference (ftp://ftp-mouse.sanger.ac.uk/ref/).

<<eval=FALSE, echo=TRUE>>=

library(oligoMask)
library(BSgenome.Mmusculus.GRC.m38)

vcf.files <- "/Users/bottomly/Desktop/resources/vcfs/mgp.v3.snps.rsIDdbSNPv137.vcf.gz"
vcf.labels <- "SNV"
#the paste procedure is to prevent running off the margin
probe.tab.file <- paste("/Users/bottomly/Desktop/resources/microarray/MoGene-1_0-st-v1.probe.tab/",
	"MoGene-1_0-st-v1.probe.tab")
strain.names <- "NODShiLtJ"
bs.genome <-BSgenome.Mmusculus.GRC.m38
db.schema <- SangerTableSchemaList()
db.name="om.NOD.mogene.1.0.st"
    
keep.category <- "main"
window.size <- 1000
max.mismatch <- 1
should.debug <- TRUE

limit.chr <- c(1:19, "X", "Y", "MT")

 package.info <- list(AUTHOR="Daniel Bottomly", AUTHOREMAIL="bottomly@ohsu.edu",
	BOWTIE_PATH="/Users/bottomly/Desktop/github_projects/bowtie_build/bowtie",
        GENOME_PATH="/Users/bottomly/Desktop/resources/sequences/GRCm38_68",
	VCF_QUERY_CMD="htscmd vcfquery", VCF_TYPE="NOD")

create.sanger.mouse.vcf.db(vcf.files, vcf.labels, probe.tab.file, strain.names, bs.genome, db.schema,
	db.name, keep.category, window.size, max.mismatch, limit.chr, should.debug, package.info)

install.packages("om.NOD.mogene.1.0.st", repos=NULL, type="source")

@

The resulting package \texttt{om.NOD.mogene.1.0.st} can be built, checked and installed using the standard R facilities.  Note that running the tests using R check will only work if BOWTIE\_PATH, GENOME\_PATH and VCF\_QUERY\_CMD are properly specified.  The package can be checked without the tests
by using the '--no-tests' argument to \texttt{R CMD check}.  For now we directly install the package.

<<eval=FALSE, echo=TRUE>>=

install.packages("om.NOD.mogene.1.0.st", repos=NULL, type="source")

@

Analysis of our subset of the Sun et al data can be carried out at the metaprobeset (core) level using both \texttt{oligo} and the \texttt{limma} packages.  Here we simply find the strain differences for the vehicle class of exposure and examine
closer those strain differences where NOD has lower expression than B6.

<<eval=TRUE, echo=FALSE>>=

load("/Users/bottomly/Desktop/github_projects/oligoMask_vignette/SunGeneFS.RData")

@

<<>>=

library(limma)
library(oligo)
library(pd.mogene.1.0.st.v1)

sum.gfs <- rma(sun.gene.fs, target="core")

sum.exprs <- exprs(sum.gfs)

phen.dta <- data.frame(t(sapply(strsplit(colnames(sum.exprs), "_"), c))[,1:3])
names(phen.dta) <- c("tissue", "strain", "exposure")

use.mod <- model.matrix(~strain, data=phen.dta)

fit <- lmFit(sum.exprs, use.mod)
fit <- eBayes(fit)

test.res <- decideTests(fit)

sum(abs(test.res[,"strainNOD"]) == 1)
sum(test.res[,"strainNOD"] == 1)
sum(test.res[,"strainNOD"] == -1)

@

To explore this further, we carry out the same analysis, except perform a masking procedure prior to the RMA preprocessing step.

<<>>=

library(om.NOD.mogene.1.0.st)s

var.parms <- VariantMaskParams(om.NOD.mogene.1.0.st, geno.filter=FALSE, rm.unmap=FALSE,
	rm.mult=FALSE)

sum.gfs.mask <- maskRMA(sun.gene.fs, target="core", apply.mask=TRUE, mask.params=var.parms)

sum.exprs.mask <- exprs(sum.gfs.mask)

stopifnot(all(colnames(sum.exprs.mask) == colnames(sum.exprs)))
#can use the same model matrix from before

fit.mask <- lmFit(sum.exprs.mask, use.mod)
fit.mask <- eBayes(fit.mask)

test.res.mask <- decideTests(fit.mask)

sum(test.res.mask[,"strainNOD"] == -1)
sum(test.res.mask[,"strainNOD"] == 1)

@

We next can explore further into why there are differences between the metaprobesets.  First we determine which metaprobesets were lost from the masking procedure.  That is, in this case, which ones had all probes impacted by a variant.  Of those
we can then determine which of those were significant in the original tests.

<<>>=

lost.mps <- setdiff(rownames(test.res), rownames(test.res.mask))

lost.mps

sig.lost.mps <- intersect(rownames(test.res)[test.res[,"strainNOD"] == -1], lost.mps)

sig.lost.mps

@

That is there are three metaprobesets that were significant in the original analysis that were removed entirely due to variants in each probe.  We can then examine these further as a sanity check using a (currently) unexported function.
First we will retrieve the necessary probes using convienience functions from the \texttt{oligo} package.  Then we will supply the set of unique probe IDs to the \texttt{getProbeVars} function from \texttt{oligoMask}.  The resulting data.frame
can then be filtered to remove all reference genotypes (\texttt{allele\_num = 0} in this case) and the variant -> probe mappings can be examined further.

<<>>=

mps.dta <- getProbeInfo(sum.gfs, probeType="pm", target="core")

var.mps <- mps.dta[mps.dta$man_fsetid %in% sig.lost.mps,]

var.probes <- unique(var.mps$fid)

var.probes.dta <- oligoMask:::getProbeVars(om.NOD.mogene.1.0.st, probe.ids=as.character(var.probes))

head(var.probes.dta)

any(var.probes.dta$allele_num %in% c(0, -1))

length(intersect(var.probes.dta$probe_id, var.mps$fid)) == length(union(var.probes.dta$probe_id,
	var.mps$fid))

@


\section{Future Directions}

Though currently we only have support for Affymetrix Gene arrays, additional Affymetrix array types supported by \texttt{oligo} can be added upon request.  We will add support for the use of the \texttt{ff} package in conjunction with \texttt{oligo} in the near future.


\section{References}

Alberts, Rudi, et al. "Sequence polymorphisms cause many false cis eQTLs." PloS one 2.7 (2007): e622.


Benovoy, David, Tony Kwan, and Jacek Majewski. "Effect of polymorphisms within probeÐtarget sequences on olignonucleotide microarray experiments." Nucleic acids research 36.13 (2008): 4417-4423.


Chia, Ruth, et al. "The origins and uses of mouse outbred stocks." Nature genetics 37.11 (2005): 1181-1186.


Collaborative Cross Consortium. "The genome architecture of the Collaborative Cross mouse genetic reference population." Genetics 190 (2012): 389-401.


Sun, Wei, et al. "Transcriptome atlases of mouse brain reveals differential expression across brain regions and genetic backgrounds." G3: Genes| Genomes| Genetics 2.2 (2012): 203-211.


Svenson, Karen L., et al. "High-resolution genetic mapping using the mouse diversity outbred population." Genetics 190.2 (2012): 437-447.


Walter, Nicole AR, et al. "SNPs matter: impact on detection of differential expression." Nature methods 4.9 (2007): 679-680.

<<eval=TRUE>>=
sessionInfo()
@

\end{document}